recipes/nasa-smap-sss/recipe.py

```diff
@@ -0,0 +1,189 @@
+    date : str
+        if freq == '8day_running':
+            and algorithm == 'JPL': 8-character numerical string of format YYYYMMDD
+            and algorithm == 'RSS': 7-character numerical string of format YYYYDDD
+        else: 6-character numerical string of format YYYYMM
+    counter_dict : dict
+        A dict of dicts.
+
+    Returns
+    -------
+    Full url path for specified dataset.
+    """
+    assert algorithm in ('JPL', 'RSS'), "algorithm must be one of 'JPL' or 'RSS'"
+    assert freq in ('8day_running', 'monthly'), "freq must be one of '8day_running' or 'monthly'"
+
+    if freq == '8day_running':
+        if algorithm == 'JPL':
+            assert len(date) == 8, "if freq=='8day_running' and algo=='JPL', len(date) must == 8"
+        elif algorithm == 'RSS':
+            assert len(date) == 7, "if freq=='8day_running' and algo=='RSS', len(date) must == 7"
+        count = counter_dict[algorithm][date]
+    else:
+        assert len(date) == 6, "if freq=='monthly', len(date) must == 6"
+        count = ''
+
+    url_base = 'https://podaac-opendap.jpl.nasa.gov/opendap/allData/smap/L3/'
+    yr = date[:4]
+
+    if algorithm == 'JPL':
+        mo = date[4:6]
+        day = date[6:8] if len(date) == 8 else None
+        freq_caps = '8DAYS' if freq == '8day_running' else 'MONTHLY'
+        file_date = f'{yr}{mo}{day}' if freq == '8day_running' else f'{yr}{mo}'
+        url_tail = f'JPL/V5.0/{freq}/{yr}/{count}/SMAP_L3_SSS_{file_date}_{freq_caps}_V5.0.nc'
+    else:
+        if len(date) == 7:
+            seq_day = date[4:7]
+            assert int(seq_day) in range(1, 367), 'Sequential day must be in range(1,367)'
+            date_spec = seq_day
+        else:
+            month = date[4:6]
+            assert int(month) in range(1, 13), 'Month must be in range(1,13)'
+            date_spec = month
+        url_tail = (
+            f'RSS/V4/{freq}/SCI/{yr}/{count}/RSS_smap_SSS_L3_{freq}_{yr}_{date_spec}_FNL_v04.0.nc'
+        )
+
+    return url_base + url_tail
+
+
+jpl_daily_dates, jpl_monthly_dates = _return_jpl_dates()
+rss_daily_dates, rss_monthly_dates = _return_rss_dates()
+
+base = 'NASA-SMAP-SSS'
+urls = {
+    f'{base}/JPL/8day': [make_path('JPL', '8day_running', date=d) for d in jpl_daily_dates],
+    f'{base}/JPL/monthly': [make_path('JPL', 'monthly', date=d) for d in jpl_monthly_dates],
+    f'{base}/RSS/8day': [make_path('RSS', '8day_running', date=d) for d in rss_daily_dates],
+    f'{base}/RSS/monthly': [make_path('RSS', 'monthly', date=d) for d in rss_monthly_dates],
+}
+
+patterns = {
+    list(urls)[i]: (
+        pattern_from_file_sequence(
+            file_list=urls[list(urls)[i]],
+            concat_dim='time',
+            nitems_per_file=1,
+            fsspec_open_kwargs={'block_size': 0},
+        )
+    )
+    for i in range(4)
+}
+
+recipes = {
+    list(patterns)[i]: (
+        XarrayZarrRecipe(
+            patterns[list(patterns)[i]],
+        )
+    )
+    for i in range(4)
+}
+
```

Comment:
@jbusecke, i could use your help here. i'm not very familiar with recipes that define multiple recipes within a dict, and for reason i don't understand yet, the backend seems to think this recipe contains errors that result in a infinite loop when synchronize this recipe with the database. unfortunately, the error message returned is a opaque and 



can you take a look at this, and let me know if anything stands out? is this dict well defined?  thank you

Reply:
i'm planning to post the error message here later today

Reply:
@jbusecke, here's the traceback, which seems to hint at a missing key from the recipes dict. is the multi recipes approach within the same feedstock documented somewhere? I couldn't find anything in the documentation. 

```python
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827794&selected=1522411785187827794) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827794&selected=1522411785187827794) Traceback (most recent call last):
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827795&selected=1522411785187827795) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827795&selected=1522411785187827795)   File "/usr/local/lib/python3.9/dist-packages/uvicorn/protocols/http/h11_impl.py", line 373, in run_asgi
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827796&selected=1522411785187827796) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827796&selected=1522411785187827796)     result = await app(self.scope, self.receive, self.send)
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827797&selected=1522411785187827797) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827797&selected=1522411785187827797)   File "/usr/local/lib/python3.9/dist-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827798&selected=1522411785187827798) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827798&selected=1522411785187827798)     return await self.app(scope, receive, send)
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827800&selected=1522411785187827800) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827800&selected=1522411785187827800)   File "/usr/local/lib/python3.9/dist-packages/fastapi/applications.py", line 208, in __call__
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827801&selected=1522411785187827801) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827801&selected=1522411785187827801)     await super().__call__(scope, receive, send)
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827802&selected=1522411785187827802) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827802&selected=1522411785187827802)   File "/usr/local/lib/python3.9/dist-packages/starlette/applications.py", line 112, in __call__
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827804&selected=1522411785187827804) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827804&selected=1522411785187827804)     await self.middleware_stack(scope, receive, send)
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827805&selected=1522411785187827805) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827805&selected=1522411785187827805)   File "/usr/local/lib/python3.9/dist-packages/starlette/middleware/errors.py", line 181, in __call__
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827808&selected=1522411785187827808) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827808&selected=1522411785187827808)     raise exc
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827810&selected=1522411785187827810) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827810&selected=1522411785187827810)   File "/usr/local/lib/python3.9/dist-packages/starlette/middleware/errors.py", line 159, in __call__
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827811&selected=1522411785187827811) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827811&selected=1522411785187827811)     await self.app(scope, receive, _send)
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827812&selected=1522411785187827812) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827812&selected=1522411785187827812)   File "/usr/local/lib/python3.9/dist-packages/starlette/middleware/cors.py", line 84, in __call__
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827813&selected=1522411785187827813) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827813&selected=1522411785187827813)     await self.app(scope, receive, send)
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827814&selected=1522411785187827814) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827814&selected=1522411785187827814)   File "/usr/local/lib/python3.9/dist-packages/starlette/exceptions.py", line 82, in __call__
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827815&selected=1522411785187827815) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827815&selected=1522411785187827815)     raise exc
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827816&selected=1522411785187827816) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827816&selected=1522411785187827816)   File "/usr/local/lib/python3.9/dist-packages/starlette/exceptions.py", line 71, in __call__
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827817&selected=1522411785187827817) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827817&selected=1522411785187827817)     await self.app(scope, receive, sender)
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827818&selected=1522411785187827818) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827818&selected=1522411785187827818)   File "/usr/local/lib/python3.9/dist-packages/starlette/routing.py", line 656, in __call__
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827819&selected=1522411785187827819) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827819&selected=1522411785187827819)     await route.handle(scope, receive, send)
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827820&selected=1522411785187827820) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827820&selected=1522411785187827820)   File "/usr/local/lib/python3.9/dist-packages/starlette/routing.py", line 259, in handle
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827821&selected=1522411785187827821) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827821&selected=1522411785187827821)     await self.app(scope, receive, send)
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827822&selected=1522411785187827822) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827822&selected=1522411785187827822)   File "/usr/local/lib/python3.9/dist-packages/starlette/routing.py", line 64, in app
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827823&selected=1522411785187827823) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827823&selected=1522411785187827823)     await response(scope, receive, send)
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827825&selected=1522411785187827825) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827825&selected=1522411785187827825)   File "/usr/local/lib/python3.9/dist-packages/starlette/responses.py", line 159, in __call__
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827826&selected=1522411785187827826) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827826&selected=1522411785187827826)     await self.background()
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827828&selected=1522411785187827828) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827828&selected=1522411785187827828)   File "/usr/local/lib/python3.9/dist-packages/starlette/background.py", line 35, in __call__
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827830&selected=1522411785187827830) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827830&selected=1522411785187827830)     await task()
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827832&selected=1522411785187827832) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827832&selected=1522411785187827832)   File "/usr/local/lib/python3.9/dist-packages/starlette/background.py", line 18, in __call__
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827833&selected=1522411785187827833) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827833&selected=1522411785187827833)     await self.func(*self.args, **self.kwargs)
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827834&selected=1522411785187827834) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827834&selected=1522411785187827834)   File "/opt/app/pangeo_forge_orchestrator/routers/github_app.py", line 783, in synchronize
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827835&selected=1522411785187827835) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827835&selected=1522411785187827835)     new_models = [
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827836&selected=1522411785187827836) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827836&selected=1522411785187827836)   File "/opt/app/pangeo_forge_orchestrator/routers/github_app.py", line 785, in <listcomp>
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827837&selected=1522411785187827837) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827837&selected=1522411785187827837)     recipe_id=recipe["id"],
Oct 27 15:07:57 [pangeo-forge-api-prod](https://my.papertrailapp.com/systems/pangeo-forge-api-prod/events?focus=1522411785187827838&selected=1522411785187827838) [app/web.1](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.1&focus=1522411785187827838&selected=1522411785187827838) KeyError: 'id'
```